### PR TITLE
Add +Use Ship as Template

### PR DESCRIPTION
**Pull request by BMagnu**
Allows to table sort of derived ships.
Behaves pretty much identical to templates (which should be preferred), but allows to use ships that may be present in a parent mod or are already tabled as templateless ships to be used as such.

### CODE REVIEW

File: code/ship/ship.cpp [hunk 1]
```diff
@@ -2228,12 +2228,32 @@ static void parse_ship(const char *filename, bool replace)
...
+				create_if_not_found = false;
 			}
 			else {
 				Warning(LOCATION, "Unable to find ship template '%s' requested by ship class '%s', ignoring template request...", template_name, fname);
 			}
 		}
 	}
+	if( optional_string( "+Use Ship as Template:" ) ) {
+		if ( !create_if_not_found ) {
+			Warning(LOCATION, "Both '+nocreate' and '+Use Ship as Template:' were specified for ship class '%s', ignoring '+Use Ship as Template:'\n", fname);
```
Comment (MageKing17) [inline code/ship/ship.cpp:2240]: This conditional could also be reached if `+Use Template:` and `+Use Ship as Template:` were used in the same entry, so this warning might not be accurate.

**Review by MageKing17 — COMMENTED** [review]
There's a reason Variants aren't already in the codebase; namely, what happens if the ship you're using as a "template" is modified by a modular table parsed after this one?

File: code/ship/ship.cpp (hunk 1 above)
Comment (BMagnu) [reply code/ship/ship.cpp:2240]: Fair point on the warning.
As for the modification of a ship after a template is defined:
I want to warn about this exact possibility in the documentation on the wiki. It is _explicitly_ possible to modify a ship after cloning it. This is intended as an expert option, for situations where you need to clone from upstream mods, whereas normal users are supposed to use templates anyways.
If _really_ necessary, I could add a Ship-Info flag that warns about modifying a cloned ship later down the line unless explicitly flagged as intended when modifying the ship, but I think that's overkill

Comment (MageKing17) [reply code/ship/ship.cpp:2240]: > If _really_ necessary, I could add a Ship-Info flag that warns about modifying a cloned ship later down the line unless explicitly flagged as intended when modifying the ship, but I think that's overkill

I agree; that sounds like overkill. As long as the documentation for the feature explicitly points out that any changes to the base ship class in later-parsed entries won't be reflected in ship classes that used them as a template, I think adding this as an option should be fine.

**Review by MageKing17 — APPROVED** [review]
Well, I don't like the amount of duplicated code (offends my DRY sensibilities), but not enough to hold up this PR.

**Pull request merged**